File: FastSort.py
#!/usr/bin/env python
#!coding=utf-8
__author__ = 'zhoupan'
def Sort(list,start,end):
    #如果元素只有一个，则直接返回
    i=start
    j=end
    k=list[start]
    while(i<j):
        while(i<j and list[j] >= k):
            j-=1
        list[i]=list[j]
        while(i<j and list[i] < k):
            i+=1
        list[j]=list[i]
        list[i]=k
    return i
def FastSort(list,start,end):
    if(start<end):
        t=Sort(list,start,end)
        FastSort(list,start,t-1)
        FastSort(list,t+1,end)
if __name__ == '__main__':
    list=[5,8,3,6,4,2]
    FastSort(list,0,len(list)-1)
    print(list)

File: FastSort.c
#include<stdio.h>
int sort(int a[],int start,int end)
{
    int i,j,k=a[start];
    i=start;
    j=end;
    while(i<j)
    {
        while(i<j && a[j] >= k)
        {
            j-=1;
        }
        a[i]=a[j];
        while(i<j && a[i] <= k)
        {
            i+=1;
        }
        a[j]=a[i];
        a[i]=k;
    }
    return i;
}
void fastsort(int a[],int start,int end)
{
    if(start<end)
    {
        int middle = sort(a,start,end);
        fastsort(a,start,middle-1);
        fastsort(a,middle+1,end);
    }
}
int main()
{
    int i,a[]={2,4,6,3,1,5};
    fastsort(a,0,5);
    for(i=0;i<6;i++)
    {
        printf("%d\t",a[i]);
    }
    printf("\n");
}
